k2/csrc/pinned_context.cu

```diff
@@ -0,0 +1,357 @@
+/**
+ * @brief PinnedContext that allocates pinned memory.
+ *
+ * The implementation is modified from
+ * https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCCachingHostAllocator.cpp
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#include <deque>
+#include <functional>
+#include <mutex>  // NOLINT
+#include <set>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "k2/csrc/context.h"
+#include "k2/csrc/log.h"
+#include "k2/csrc/macros.h"
+#include "k2/csrc/nvtx.h"
+
+namespace k2 {
+
+namespace {
+
+struct BlockSize {
+  size_t size;  // size of this memory block in bytes
+  void *ptr;    // pointer to the beginning of this memory block
+
+  explicit BlockSize(size_t size, void *ptr = nullptr) : size(size), ptr(ptr) {}
+};
+
+struct Block : public BlockSize {
+  bool allocated;  // true if the block is currently allocated
+                   // false if the block is available for allocation
+
+  int event_count;  // number of outstanding cuda events
+  std::unordered_set<cudaStream_t> streams;
+
+  Block(size_t size, void *ptr, bool allocated)
+      : BlockSize(size, ptr), allocated(allocated), event_count(0), streams() {}
+};
+
+static bool BlockComparator(const BlockSize &a, const BlockSize &b) {
+  NVTX_RANGE(K2_FUNC);
+  // sort by size, break ties with pointer
+  if (a.size != b.size) return a.size < b.size;
+
+  return std::less<void *>()(a.ptr, b.ptr);
+}
+
+/* Allocate pinned memory using cudaMallocHost with caching.
+
+  WARNING: Once memory is allocated, it is not returned to the system.
+  Use it with care!
+ */
+class PinnedAllocator {
+ public:
+  PinnedAllocator() : available_(&BlockComparator) {}
+
+  /* Allocate a block of memory.
+
+     If we can find a free block that is large enough (first fit or best fit
+     as free blocks are sorted by size) for the requested size, the free block
+     is marked as allocated and returned to the user.
+
+     If no free blocks are available, a new block is allocated by
+     using `cudaMallocHost`.
+
+     @param  [in]   size        Number of bytes to be allocated.
+     @param  [out]  ptr         On return, it contains the starting address of
+                                the allocated memory.
+
+     @return Return cudaSuccess on success. Return a CUDA error code on failure.
+   */
+  cudaError_t Malloc(size_t size, void **ptr) {
+    NVTX_RANGE(K2_FUNC);
+    K2_CHECK_NE(ptr, nullptr);
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    // If the number of outstanding cuda events is larger
+    // than 100, we will invoke `ProcessEvents()`, which
+    // may make some blocks available for reuse.
+    //
+    // If the number of cuda_events_ is small,
+    // we first try to find a block from the pool. If it
+    // cannot find one, we then invoke `ProcessEvents`.
```

Comment:
`If it cannot find one, we then invoke ProcessEvents or create a new one?` seems there's a mismatch between code and documentation?

Reply:
The comment is for the `for` loop. It does not create a new block in the `for` loop.
The `for` loop will try to reuse an existing block. But if it cannot find one, we just exit the `for` loop.
Statements after the for loop will allocate a new block.